ride/lp_stable.ride

```diff
@@ -919,6 +919,143 @@ func mm(i: Invocation) = {
+  let poolConfig = gpc()
+  let amId = poolConfig[idxAmAsId]
+  let prId = poolConfig[idxPrAsId]
+  let n = big2
+  let aPrecision = Amult.parseBigIntValue()
+  let a = A.parseBigIntValue() * aPrecision
+  let xp = if (isReverse == false) then { 
+    [amId.getAccBalance().toBigInt() + poolAmountInBalance, prId.getAccBalance().toBigInt()]
+  } else {
+    [prId.getAccBalance().toBigInt() + poolAmountInBalance, amId.getAccBalance().toBigInt()]
+  }
+
+  let x = xp[0]
+  let s = x
+  let ann = a * n
+
+  let c = D * D / (x * n) * D * aPrecision / (ann * n)
+  let b = s + D * aPrecision / ann - D
+  func calc(acc: (BigInt, Int|Unit), cur: Int) = {
+      let (y, found) = acc
+      if (found != unit) then acc else {
+        let yNext = (y * y + c) / (big2 * y + b)
+        let yDiff = absBigInt(yNext - y.value())
+        if (yDiff <= big1) then {
+          (yNext, cur)
+        } else {
+          (yNext, unit)
+        }
+      }
+    }
+    let arr = [0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14]
+    let (y, found) = FOLD<15>(arr, (D, unit), calc)
+    if (found != unit) then {
+      y
+    } else {
+      { "Y calculation error, Y = " + y.toString() }.throw()
+    }
+}
+
+@Callable(i)
+func calculateAmountOutForSwapREADONLY(cleanAmountIn: Int, isReverse: Boolean) = {
+   let (assetOut, poolAmountInBalance) = if (isReverse == false) then {
+    let assetOut = this.strf(pa())
+    let poolAmountInBalance = getAccBalance(this.strf(aa())).toBigInt() + cleanAmountIn.toBigInt()
+    (assetOut, poolAmountInBalance)
+  } else {
+    let assetOut = this.strf(aa())
+    let poolAmountInBalance = getAccBalance(this.strf(pa())).toBigInt() + cleanAmountIn.toBigInt()
+    (assetOut, poolAmountInBalance) 
+  }
+
+  let poolConfig = gpc()
+  let amId = poolConfig[idxAmAsId]
+  let prId = poolConfig[idxPrAsId]
+  let xp = [amId.getAccBalance().toBigInt(), prId.getAccBalance().toBigInt()]
+  
+  let D = xp.getD()
+  let y = getY(isReverse, D, cleanAmountIn.toBigInt())
+
+  let dy = getAccBalance(assetOut).toBigInt() - y - toBigInt(1) # -1 just in case there were some rounding errors
+
+  let totalGetRaw = [0, dy.toInt()].max()
+
+  let newXp = if (isReverse == false) then {
+    [amId.getAccBalance().toBigInt() + cleanAmountIn.toBigInt(), prId.getAccBalance().toBigInt() - dy]
+  } else { 
+    [amId.getAccBalance().toBigInt() - dy, prId.getAccBalance().toBigInt() + cleanAmountIn.toBigInt()]
+  }
+  let newD = newXp.getD()
+  strict checkD = newD >= D || makeString(["new D is fewer error", D.toString(), newD.toString()], "__").throw()
+  
+  (nil, totalGetRaw)
+}
+
+@Callable(i)
+func calculateAmountOutForSwapAndSendTokens(cleanAmountIn: Int, isReverse: Boolean, amountOutMin: Int, addressTo: String) = {
+  let swapContact = invoke(
+    fca,
+    "getSwapContractREADONLY", 
+    [], 
+    []
+  ).exactAs[String]
+  strict checks = [
+    i.payments[0].value().amount >= cleanAmountIn || "Wrong amount".throwErr(),
```

Comment:
Looks strange. `cleanAmountIn` should be equal to `i.payments[0].value().amount` so the argument `cleanAmountIn` does not needed at all.

Reply:
I use >= in case if we going to change attached payment with fee included